pkg/postgresql/postgresql.go

```diff
@@ -157,24 +165,62 @@ out:
 	return nil
 }
 
+// Start starts the PostgreSQL server
 func (p *Manager) Start() error {
 	log.Infof("Starting database")
 	if err := p.WriteConf(); err != nil {
-		return fmt.Errorf("error writing conf file: %v", err)
+		return trace.Wrap(err, "error writing conf file")
 	}
 	name := filepath.Join(p.pgBinPath, "pg_ctl")
 	cmd := exec.Command(name, "start", "-w", "-D", p.dataDir)
 	cmd.Stdout = os.Stdout
 	cmd.Stderr = os.Stderr
-	// TODO(sgotti) attaching a pipe to sdtout/stderr makes the postgres
-	// process executed by pg_ctl inheriting it's file descriptors. So
-	// cmd.Wait() will block and waiting on them to be closed (will happend
-	// only when postgres is stopped). So this functions will never return.
-	// To avoid this no output is captured. If needed there's the need to
-	// find a way to get the output whitout blocking.
-	if err := cmd.Run(); err != nil {
-		return fmt.Errorf("error: %v", err)
+
+	if err := cmd.Start(); err != nil {
+		return trace.Wrap(err, "error starting PostgreSQL process")
 	}
+
+	exited := make(chan struct{})
+	go func() {
+		_ = cmd.Wait()
+		close(exited)
+	}()
+
+	pid := cmd.Process.Pid
```

Comment:
I'd move the rest to a separate function (i.e. `waitForPostgresSQLProcess` or something similar) (see [here](http://geekswithblogs.net/theArchitectsNapkin/archive/2015/05/12/actors-in-a-ioda-architecture-by-example.aspx) and [here](http://geekswithblogs.net/theArchitectsNapkin/archive/2015/04/29/the-ioda-architecture.aspx) for inspiration).

I'd like to keep integration functions (i.e. those that call into other integrations or operations) but do not do implement non-trivial logic themselves (i.e. do not mix calling high-level APIs and complex logic like the waiting loop here) separate from operations (i.e. functional units that focus on some piece of data or I/O) - then it becomes easier to test - integrations do not require any testing since they are soley comprised of invocations to other functions (other integrations or operations) and hence are tested by definition provided that the operations are tested and operations that focus on a single task (and do not invoke other operations) are much easier to test.
This is all ideally of course - sometimes it is either not possible or feasible but striving for such separation is always beneficial in my opinion as it really helps readability.